libs/sync/operators.go

```diff
@@ -0,0 +1,171 @@
+package sync
+
+import (
+	"path"
+	"path/filepath"
+
+	"golang.org/x/exp/maps"
+)
+
+// List of operations to apply to synchronize local file systems changes to WSFS.
+type operators struct {
+	delete []string
+	rmdir  []string
+	mkdir  []string
+	put    []string
+}
+
+func (d operators) IsEmpty() bool {
+	return len(d.put) == 0 && len(d.delete) == 0
+}
+
+// Compute operations required to make files in WSFS reflect current local files.
+// Takes into account changes since the last sync iteration.
+func computeOperators(target *FilesState, current *FilesState) operators {
+	d := &operators{
+		delete: make([]string, 0),
+		rmdir:  make([]string, 0),
+		mkdir:  make([]string, 0),
+		put:    make([]string, 0),
+	}
+	d.caseFilesRemoved(target, current)
+	d.caseRemoteNameChanged(target, current)
+	d.caseFilesAdded(target, current)
+	d.caseFilesUpdated(target, current)
+	return *d
+}
+
+// Add operators for tracked files that no longer exist.
+func (d *operators) caseFilesRemoved(after *FilesState, before *FilesState) {
+	for localName, remoteName := range before.LocalToRemoteNames {
+		if _, ok := after.LocalToRemoteNames[localName]; !ok {
+			d.delete = append(d.delete, remoteName)
+		}
+	}
+
+	// Remove directories that would no longer contain any files.
+	beforeDirs := MakeDirSet(maps.Keys(before.LocalToRemoteNames))
+	afterDirs := MakeDirSet(maps.Keys(after.LocalToRemoteNames))
+	d.rmdir = append(d.rmdir, beforeDirs.Remove(afterDirs).Slice()...)
+}
+
+// Cleanup previous remote files for files that had their remote targets change. For
+// example this is possible if you convert a normal python script to a notebook.
+func (d *operators) caseRemoteNameChanged(after *FilesState, before *FilesState) {
+	for localName, beforeRemoteName := range before.LocalToRemoteNames {
+		afterRemoteName, ok := after.LocalToRemoteNames[localName]
+		if !ok || afterRemoteName == beforeRemoteName {
```

Comment:
making it `ok && afterRemoteName != beforeRemoteName` reads better from intentions perspective

Reply:
done